### DISCUSSION TITLE
Github Action build of WASM project timeout

### DISCUSSION THREAD

**Discussion by baskren** [unoplatform/uno#12669] [Q&A] [answered]
### Current behavior

Building WASM project using GitHub Action times out during download of `dotnet-runtime-wasm-linux-585dcd7-7db1c333330-4019238418-Release`

```
  Downloading dotnet-runtime-wasm-linux-585dcd7-7db1c333330-4019238418-Release to /tmp/dotnet-runtime-wasm-linux-585dcd7-7db1c333330-4019238418-Release.zip
MSBUILD : error MSB4166: Child node "2" exited prematurely. Shutting down. Diagnostic information may be found in files in "/tmp/MSBuildTemprunner/" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.
```

This used to happen once in blue moon but, lately it's almost every build.  FWIW, I doubled the `time-out:` parameter of my job to 60 minutes with no impact upon this issue.

### Expected behavior

Github Action does not time-out during download of `dotnet-runtime-wasm-linux-585dcd7-7db1c333330-4019238418-Release`

### How to reproduce it (as minimally and precisely as possible)

Share with me your GitHub account and I'll give you access to the private repo in question.

### Workaround

Love to know of one!

### Works on UWP/WinUI

None

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

    <PackageVersion Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
    <PackageVersion Include="Microsoft.Windows.Compatibility" Version="7.0.0" />
    <PackageVersion Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.756" />
    <PackageVersion Include="Microsoft.WindowsAppSDK" Version="1.3.230502000" />
    
        
    <PackageVersion Include="SkiaSharp.Skottie" Version="2.88.3" />
    <PackageVersion Include="SkiaSharp.Views.Uno.WinUI" Version="2.88.3" />
    <PackageVersion Include="Swashbuckle.AspNetCore" Version="6.5.0" />
    <PackageVersion Include="Uno.Core.Extensions.Logging.Singleton" Version="4.0.1" />
    <PackageVersion Include="Uno.Extensions.Logging.OSLog" Version="1.4.0" />
    <PackageVersion Include="Uno.Extensions.Logging.WebAssembly.Console" Version="1.4.0" />
    <PackageVersion Include="Uno.Resizetizer" Version="1.1.0" />
    <PackageVersion Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="4.9.17" />
    <PackageVersion Include="Uno.UniversalImageLoader" Version="1.9.36" />
    <PackageVersion Include="Uno.Wasm.Bootstrap" Version="7.0.23" />
    <PackageVersion Include="Uno.Wasm.Bootstrap.DevServer" Version="7.0.23" />
    <PackageVersion Include="Uno.Wasm.Bootstrap.Server" Version="7.0.23" />
    <PackageVersion Include="Uno.WinUI" Version="4.9.17" />
    <PackageVersion Include="Uno.WinUI.RemoteControl" Version="4.9.17" />
    <PackageVersion Include="Uno.WinUI.Skia.Wpf" Version="4.9.17" />
    <PackageVersion Include="Uno.WinUI.WebAssembly" Version="4.9.17" />
    <PackageVersion Include="Xamarin.Google.Android.Material" Version="1.8.0" />

    <PackageVersion Include="Uno.WinUI.Lottie" Version="4.9.17" />
    <PackageVersion Include="Microsoft.SourceLink.GitHub" Version="1.1.1"/>
	<PackageVersion Include="Newtonsoft.Json" Version="13.0.2" />

		<PackageVersion Include="SkiaSharp.Svg" Version="1.60.0" />

		<PackageVersion Include="Microsoft.AppCenter.Distribute" Version="5.0.1" />
		<PackageVersion Include="Microsoft.AppCenter.Analytics" Version="5.0.1" />
		<PackageVersion Include="Microsoft.AppCenter.Crashes" Version="5.0.1" />

### Affected platforms

WebAssembly

### IDE

Visual Studio 2022

### IDE version

Version 17.7.0 Preview 2.0

**Comment by jeromelaban** [comment]
Thanks for the report! This looks like a memory issue, where the github actions agents are limited and WebAssembly builds (particularly with AOT enabled) can consume a lot of memory.

There are two changes that you can make:
- Use profiled AOT so that you get less code compiled to AOT, consuming less memory
- Reduce the linker impact by [using these flags](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-linker-opts.html#emscripten-linker-optimizations-flags). The `WasmShellEmccLinkOptimization` option may be the most efficient to get you unblocked.

**Comment by baskren** [comment, marked as answer]
Hi @jeromelaban ,

Thanks for the ideas.  Now that I've had a chance to try a few things, I wanted to close the loop with you on this and share what I learned.  The app was created using the default `<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>` so your first suggestion was already on my "things to try" list.

1. **Profiled AOT**: [Followed these instructions](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/runtime-execution-modes.html#profile-guided-aot) but got the following error in the JavaScript console that coincided with some functionality of the app not working:
```
Error: [MONO] * Assertion at D:/a/Uno.DotnetRuntime.WebAssembly/Uno.DotnetRuntime.WebAssembly/runtime/src/mono/mono/mini/interp/interp.c:3125, condition `is_ok (error)' not met, function:interp_create_method_pointer_llvmonly, couldn't compile wrapper "(wrapper other) intptr object:interp_in_generic (intptr,intptr,intptr,intptr,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,intptr,intptr)" for "System.Threading.Tasks.Task`1<P42.Uno.Controls.PermissionPopup> P42.Uno.Controls.PermissionPopup:CreateAsync (string,string,string,string,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,P42.Uno.Controls.Effect,P42.Uno.Controls.EffectMode)", due to Attempting to JIT compile method '(wrapper other) intptr object:interp_in_generic (intptr,intptr,intptr,intptr,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,Windows.UI.Color,intptr,intptr)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
 assembly:<unknown assembly> type:<unknown type> member:(null)

    at mono_wasm_stringify_as_error_with_stack (dotnet.js:5:620)
    at Object.mono_wasm_trace_logger (dotnet.js:5:985)
    at _mono_wasm_trace_logger (dotnet.js:14:140504)
    at dotnet.wasm:0x10ae6dc
    at dotnet.wasm:0x1128b8a
    at dotnet.wasm:0x83d1e8
    at dotnet.wasm:0x2e16a
    at dotnet.wasm:0x88cf2e
    at dotnet.wasm:0x1c0683
    at dotnet.wasm:0x26cf36 ...
```

  My first guess was that this could be addressed by applying [AOT Profile method exclusion](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/runtime-execution-modes.html#aot-profile-method-exclusion) but I haven't taken the time yet to dive into that.  I have some doubts that my guess was right because I was using `<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>`and seeing the following in the above error message:
 - `... couldn't compile wrapper ...`
 -  `... due to Attempting to JIT compile method ...` 
 - `... while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations ` 

2. **Reduce linker impact**: The day that I was working on this, for some reason I couldn't get [this web page](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-linker-opts.html#emscripten-linker-optimizations-flags) to open.  Loads fine now!  That, and my doubts about getting Item 1, above, to work without a lot of dead-ends (which, given the build times, would consume a lot of time), I decided to see if I could figure out a way to build locally and upload the results directly to my Azure Static WebApp. 

3. **Build via GitHub Actions Local Runner**:  My first attempt at building locally was to try out GitHub Actions Local Runner.  I tried local runners in an Ubuntu Arm64 VM, Locally in MacOS, and in a Windows 11 ARM64 VM on my M2 Mac:

  - Ubuntu ARM64 and MacOS M2: Emscripten failed to install via Uno.Wasm.Bootstrap.targets:

  ```
/home/parallels/.nuget/packages/uno.wasm.bootstrap/7.0.23/build/Uno.Wasm.Bootstrap.targets(207,3): error : System.NotSupportedException: Failed to install emscripten [/home/parallels/actions-runner/_work/HeightsAndAreas.WinUI/HeightsAndAreas.WinUI/HeightsAndAreas.Wasm/HeightsAndAreas.Wasm.csproj]
/home/parallels/.nuget/packages/uno.wasm.bootstrap/7.0.23/build/Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_vebfec278e3a4de2a23ab13f05c05a55fcb01fe1e.ValidateEmscripten() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 1269 [/home/parallels/actions-runner/_work/HeightsAndAreas.WinUI/HeightsAndAreas.WinUI/HeightsAndAreas.Wasm/HeightsAndAreas.Wasm.csproj]
/home/parallels/.nuget/packages/uno.wasm.bootstrap/7.0.23/build/Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_vebfec278e3a4de2a23ab13f05c05a55fcb01fe1e.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 769 [/home/parallels/actions-runner/_work/HeightsAndAreas.WinUI/HeightsAndAreas.WinUI/HeightsAndAreas.Wasm/HeightsAndAreas.Wasm.csproj]
/home/parallels/.nuget/packages/uno.wasm.bootstrap/7.0.23/build/Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_vebfec278e3a4de2a23ab13f05c05a55fcb01fe1e.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 260 [/home/parallels/actions-runner/_work/HeightsAndAreas.WinUI/HeightsAndAreas.WinUI/HeightsAndAreas.Wasm/HeightsAndAreas.Wasm.csproj]
  ```

  - Windows 11 ARM64:  Build was successful (no issues loading Emscripten) but deployment to Azure failed:

  ```
Run Azure/static-web-apps-deploy@v1
  with:
    azure_static_web_apps_api_token: ***
    repo_token: ***
    action: upload
    app_location: HeightsAndAreas.Wasm/bin/Release/net7.0/dist
  env:
    P42_TargetOnlyThisPlatform: wasm
    DOTNET_ROOT: C:\Users\ben\AppData\Local\Microsoft\dotnet
Error: Container action is only supported on Linux
  ```

  So, for my M2 Mac, I was out of options for getting GitHub Actions Local Runner to work!

4. **Azure Static Web Apps CLI**: Somewhere in trying to get my local runner to work, I discovered that [you can deploy directly to an Azure Static Web App the build results!](https://learn.microsoft.com/en-us/azure/static-web-apps/static-web-apps-cli-deploy)  If I remember right, I had to configure the Static Web App from the beginning for this to work but don't quote me on that.

So, my work around turned out be deploying the Release Build that I tested on my local machine directly to Azure using the Static Web Apps CLI.

Again, thanks for your willingness to share some ideas with me.

**Reply by jeromelaban** [reply]
Thanks for the details. A few things:

- > Ubuntu ARM64 and MacOS M2: Emscripten failed to install via Uno.Wasm.Bootstrap.targets:

It may be that emscripten is not yet built to run on arm64 for Linux. You can try opening a binlog and check the error messages to see what could be the actual cause of the error.

- Profiled AOT errors may happen sometimes, depending on how the profile has been built. .NET 8 has some fixes in that area, with some more coming up, but that's correct, using method exclusions is required in those cases. Errors reported, along with their stack traces, may help in finding which methods need to be excluded.

- You should be able to build using azure static apps as a target, and github actions just fine. You take a look at [this repo](https://github.com/unoplatform/Uno.QuickStart/blob/main/.github/workflows/dotnet.yml) for reference.

**Reply by baskren** [reply]
Thanks, @jeromelaban!  When I have some free time, I'll give method exclusions a try. 

Also, FWIW, it also occurred to me that emcripten might not be available for MacOS M2 or Ubuntu ARM64 so I tried installing it directly using brew and apt, respectively, in both environments.  Both were successful.

Again, Thanks for the feedback!
